docs/source/type_narrowing.rst

```diff
@@ -0,0 +1,323 @@
+**************
+Type Narrowing
+**************
+
+Python programs often contain symbols that take on multiple types within a
+single given scope and that are distinguished by a conditional check at
+runtime. For example, here the variable *name* can be either a ``str`` or
+``None``, and the ``if name is not None`` narrows it down to just ``str``::
+
+    def maybe_greet(name: str | None) -> None:
+        if name is not None:
+            print("Hello, " + name)
+
+This technique is called *type narrowing*.
+To avoid false positives on such code, type checkers understand
+various kinds of conditional checks that are used to narrow types in Python code.
+The exact set of type narrowing constructs that a type checker understands
+is not specified and varies across type checkers. Commonly understood
```

Comment:
Should we have a minimum required set of narrowing constructs? The one bellow might as well be one with a few additions, e.g. `issubclass`.

Reply:
> Should we have a minimum required set of narrowing constructs?

That would require a change to the spec, which would have to go through the typing council. This PR doesn't propose a change to the spec, it just adds a new page of user-facing documentation (it's descriptive, not prescriptive).

Reply:
Sure, I guess this is a request for a spec update then :)

Reply:
Covering this behavior in the spec is probably useful but it's a much larger project. This PR is meant as user-facing documentation, not a full spec.